### PR TITLE
add addresses.py and test_addresses.py

### PR DESCRIPTION
**Pull request by mdesjardins**
# Code Review Activity

This repository contains a modified code sample from Sendwave's actual codebase. Please **submit a code review, including suggestions for improving code quality.** Assume your "audience" for this code review is a less experienced engineer; try to communicate your feedback clearly and compassionately.

You may also reference the [Address model definition ](https://github.com/waveremit/code-review-activity/blob/master/src/models/address.py) for context.

This activity should take about **30 minutes.**

**How to Submit A Code Review Via Github**:
* Click the "Files changed" tab above
* Click on a part of the code you want to comment on; write your comment in the input box and click "Start Review"
* Add more comments by clicking the code, writing your comment, and clicking "Add Review Comment"
* Click the green "Finish Your Review" button at the top of the page, write a final comment, and submit

### CODE REVIEW

File: src/addresses.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        return {}
+    return {}
+
+def get_city_and_state_from_postal_code(
+    country_name: str, # the two letter code for the country
+    postal_code: str
+) -> Tuple[Optional[str], Optional[str]]:
+    """ Given a country code and a zip, fetch the city and state."""
+    standardized_country_code = country_name.upper()
+    if standardized_country_code == 'GB' or standardized_country_code == 'UK':
```
Comment (akandada) [inline src/addresses.py]: if standardized_country_code in ['GB', 'UK']:

File: src/addresses.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        city = city,
+        state = state,
+        postal_code = postal_code,
+        postal_code_city = postal_code_city,
+        postal_code_state = postal_code_state,
+        country=country.lower()
+    )
+
+def get_api_headers(api_name: str) -> Dict:
+    if api_name == 'ziptastic':
```
Comment (akandada) [inline src/addresses.py]: if api_name == 'ziptastic':
    return {"x-key": config.ZIPTASTIC_API_KEY}
else:
   return {}

File: src/addresses.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        return (city, None)
+    elif (
+        standardized_country_code == 'US'
+        or standardized_country_code == 'CA'
+        or standardized_country_code == 'fr'
+    ):
+        # FYI: Other countries Ziptastic supports:
+        # https://www.getziptastic.com/faq
+        try:
+            resp = requests.get(
```
Comment (akandada) [inline src/addresses.py]: resp = requests.get(f'{BASE_ZIPTASTIC_API_URL}{postal_code}',
--
  | headers=get_api_headers('postcodes'))

File: src/addresses.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+def get_city_and_state_from_postal_code(
+    country_name: str, # the two letter code for the country
+    postal_code: str
+) -> Tuple[Optional[str], Optional[str]]:
+    """ Given a country code and a zip, fetch the city and state."""
+    standardized_country_code = country_name.upper()
+    if standardized_country_code == 'GB' or standardized_country_code == 'UK':
+        resp = requests.get(
+            BASE_POSTCODES_API_URL+postal_code,
```
Comment (akandada) [inline src/addresses.py]: resp = requests.get(f'{BASE_ZIPTASTIC_API_URL}{postal_code}', headers=get_api_headers('postcodes'))

File: src/addresses.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            headers = get_api_headers('postcodes')
+        )
+        rjson = resp.json()
+        result = rjson.get('result', {})
+        city = result['admin_district'] if 'admin_district' in result else None
+        return (city, None)
+    elif (
```
Comment (akandada) [inline src/addresses.py]: elif standardized_country_code in ['US', 'CA', 'fr']:

Comment (akandada) [reply src/addresses.py]: fr != FR?

File: src/addresses.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    ):
+        # FYI: Other countries Ziptastic supports:
+        # https://www.getziptastic.com/faq
+        try:
+            resp = requests.get(
+                f'{BASE_ZIPTASTIC_API_URL}{standardized_country_code}/{postal_code}',
+                headers = get_api_headers('ziptastic')
+            )
+            rjson = resp.json()
+        except:
```
Comment (akandada) [inline src/addresses.py]: except requests.exceptions.RequestException as e
Handle exception and log it

File: src/addresses.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    country: str
+) -> Address:
+    """Constructs an Address instance with properly formatted fields and, if
+    applicable:
+        * `postal_code_city`: the city inferred from the zip code
+        * `postal_code_state`: the state inferred from the zip code
+    """
+    (postal_code_city, postal_code_state
+        ) = get_city_and_state_from_postal_code(country, postal_code)
+    return Address(
```
Comment (akandada) [inline src/addresses.py]: If city and state code come back as None, should we still have the record?

File: src/addresses.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        try:
+            resp = requests.get(
+                f'{BASE_ZIPTASTIC_API_URL}{standardized_country_code}/{postal_code}',
+                headers = get_api_headers('ziptastic')
+            )
+            rjson = resp.json()
+        except:
+            return (None, None)
+
+        city = rjson.get('city', '')
```
Comment (akandada) [inline src/addresses.py]: Move this inside try block

File: tests/test_addresses.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        assert addresses.BASE_ZIPTASTIC_API_URL in path
+
+        # Check function return value
+        assert city == 'Brooklyn'
+        assert state == 'NY'
+
+    def test_get_city_and_state_unspported_country(self):
+        (city, state) = addresses.get_city_and_state_from_postal_code('gh', '11216')
+        assert city == None
+        assert state == None
```
Comment (akandada) [inline tests/test_addresses.py]: Add a test case where either City or state come back as None

File: src/addresses.py [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    standardized_country_code = country_name.upper()
+    if standardized_country_code == 'GB' or standardized_country_code == 'UK':
+        resp = requests.get(
+            BASE_POSTCODES_API_URL+postal_code,
+            headers = get_api_headers('postcodes')
+        )
+        rjson = resp.json()
+        result = rjson.get('result', {})
+        city = result['admin_district'] if 'admin_district' in result else None
+        return (city, None)
```
Comment (akandada) [inline src/addresses.py]: state missing?

File: src/addresses.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+from configvars import config
+from models import Address
+
+# deps
+import requests
+
+BASE_ZIPTASTIC_API_URL = 'http://zip.getziptastic.com/v2/'
+BASE_POSTCODES_API_URL = 'http://api.postcodes.io/postcodes/'
+
+def construct_address(
```
Comment (akandada) [inline src/addresses.py]: Address pep8 across the board  >> https://www.python.org/dev/peps/pep-0008/
